File: c++/InsertionSortList.cpp
// Sort a linked list using insertion sort.

#include <iostream>
using namespace std;

// * Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
 
ListNode* init(int a[], int n) {
    ListNode* head = new ListNode(a[0]);
    ListNode *curr = head;

    for (int i = 1; i < n; ++i) {
	ListNode* tmp = new ListNode(a[i]);
	curr->next = tmp;
	curr = curr->next;
    }
    return head;
}

void printList(ListNode *head) {
    if (head == NULL) return;
    while (head != NULL) {
	cout << head->val << endl;
	head = head->next;
    }
    cout << endl;
}

ListNode* insertionSortList(ListNode *head) {
    if (head == NULL) return head;

    ListNode *p = head->next;
    int tmp;
    ListNode *nodetmp, *prev, *curr;
    while (p) {
	prev = head;
	curr = head->next;
	while (curr && curr != p) {
	    if (p->val < prev->val) {
		tmp = prev->val;
		prev->val = p->val;
		p->val = tmp;
	    } 
	    if (p->val >= prev->val && p->val < curr->val) {
		tmp = curr->val;
		curr->val = p->val;
		p->val = tmp;
	    }
	    prev = curr;
	    curr = curr->next;
	}
	if (curr == p && prev->val > curr->val) {
	    if (prev == head) {
		tmp = prev->val;
		prev->val = curr->val;
		curr->val = tmp;
	    } else {
		nodetmp = curr;
		prev->next = curr->next;
		curr = curr->next;
		delete nodetmp;
		p = p->next;
	    }
	} else 
	    p = p->next;
	prev = head;
	curr = head->next;
    }
    return head;
}

int main() {
    int a[] = {2, 3, 4, 1};
    int n = 4;
    ListNode *head = init(a, n);
    printList(head);

    ListNode *p = insertionSortList(head);
    printList(p);
    return 0;
}


/*

  
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };

class Solution {
public:
    ListNode *insertionSortList(ListNode *head) {
            if (head == NULL) return head;

    ListNode *p = head->next;
    int tmp;
    ListNode *nodetmp, *prev, *curr;
    while (p) {
	prev = head;
	curr = head->next;
	while (curr && curr != p) {
	    if (p->val < prev->val) {
		tmp = prev->val;
		prev->val = p->val;
		p->val = tmp;
	    } 
	    if (p->val >= prev->val && p->val < curr->val) {
		tmp = curr->val;
		curr->val = p->val;
		p->val = tmp;
	    }
	    prev = curr;
	    curr = curr->next;
	}
	if (curr == p && prev->val > curr->val) {
	    if (prev == head) {
		tmp = prev->val;
		prev->val = curr->val;
		curr->val = tmp;
	    } else {
		nodetmp = curr;
		prev->next = curr->next;
		curr = curr->next;
		delete nodetmp;
		p = p->next;
	    }
	} else 
	    p = p->next;
	prev = head;
	curr = head->next;
    }
    return head;

    }
};


 */

File: java/insertionSortList.java
import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;

public class insertionSortList {
    public static class Solution {
        public class ListNode {
            int val;
            ListNode next;
            ListNode(int x) {
                val = x;
                next = null;
            }
        }

        public ListNode ListNode(int x) {
            ListNode res = new Solution.ListNode(-1);
            res.val = x;
            res.next = null;
            return res;
        }

        public ListNode insertionSortList(ListNode head) {
            if (head == null) {
                return null;
            }
            
            ListNode prev = head;
            ListNode curr = head.next;
            ListNode next = head;
            ListNode tnext = head;
            ListNode tmp = head;
            while (curr != null) {
                next = curr.next;
                if (curr.val < prev.val) {
                    prev.next = next;
                    curr.next = null;
                    if (curr.val < head.val) {
                        curr.next = head;
                        head = curr;
                    } else {
                        tmp = head;
                        while (tmp.next != prev && curr.val > tmp.next.val) {
                            tmp = tmp.next;
                        }
                        tnext = tmp.next;
                        tmp.next = curr;
                        curr.next = tnext;
                    }
                    curr = next;
                } else {
                    prev = curr;
                    curr = curr.next;
                }
            }
            return head;
        }
    }

    public static void main(String[] args){
        Solution result = new Solution();
        int [] a= {6, 5, 3, 1, 2};
        Solution.ListNode one = result.ListNode(a[0]);
        Solution.ListNode curr = one;
        for (int i = 1; i < a.length; i++) {
            Solution.ListNode tmp = result.ListNode(a[i]);
            curr.next = tmp;
            curr = curr.next;
        }

        curr = result.insertionSortList(one);
        while (curr != null) {
            System.out.println(curr.val);
            curr = curr.next;
        }
    }
}
